### PR TITLE
[SPARK-27936][K8S] Support python deps

### PR DESCRIPTION
**Pull request by skonto**
### What changes were proposed in this pull request?
Supports python client deps from the launcher fs.
### Why are the changes needed?
This is a feature that was added for java deps. This PR adds support fo rpythona s well.

### Does this PR introduce any user-facing change?
yes

### How was this patch tested?
Manually running different scenarios and via examining the driver & executors logs. Also there is an integration test added.
I verified that the python resources are added to the spark file server and they are named properly so they dont fail the executors. Note here that as previously the following will not work:
primary resource `A.py`: uses a closure defined in submited pyfile `B.py`, context.py only adds to the pythonpath files with certain extension eg. zip, egg, jar.

### CODE REVIEW

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 1]
```diff
@@ -125,16 +125,16 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
 
-    kubernetesTestComponents
+    Eventually.eventually(TIMEOUT, INTERVAL) (kubernetesTestComponents
       .kubernetesClient
       .services()
-      .create(cephService)
+      .create(cephService))
 
-    kubernetesTestComponents
+    Eventually.eventually(TIMEOUT, INTERVAL) (kubernetesTestComponents
```
Comment (skonto) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: instead of watching for deletion make sure we are able to create the resources if some previous is still being deleted.

**Comment by skonto** [comment]
@erikerlandson @holdenk pls review. I know there are some other changes related to the tests, I can always rebase if they get merged first.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 2]
```diff
@@ -183,6 +185,50 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+        .set("spark.kubernetes.pyspark.pythonVersion", "2")
+        .set("spark.hadoop.fs.s3a.access.key", accessKey)
+        .set("spark.hadoop.fs.s3a.secret.key", secretKey)
+        .set("spark.hadoop.fs.s3a.connection.ssl.enabled", "false")
+        .set("spark.hadoop.fs.s3a.endpoint", s"$cephHost:$cephPort")
+        .set("spark.kubernetes.file.upload.path", s"s3a://$bucket")
+        .set("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+        .set("spark.jars.packages", "com.amazonaws:aws-java-sdk:" +
+          "1.7.4,org.apache.hadoop:hadoop-aws:2.7.6")
+        .set("spark.driver.extraJavaOptions", "-Divy.cache.dir=/tmp -Divy.home=/tmp")
```
Comment (skonto) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: I can refactor this part to have the properties set once as they are shared with the existing test. In general I think we should separate the Suites in the future to allow better setup for before and after conditions.

**Comment by holdenk** [comment]
cc @ifilonenko who I think did some related work?

**Comment by skonto** [comment]
@erikerlandson if you have some free time pls have a look :)

**Comment by erikerlandson** [comment]
@skonto this LGTM

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 3]
```diff
@@ -259,12 +259,19 @@ private[spark] object KubernetesUtils extends Logging {
...
   }
 
-  def renameMainAppResource(resource: String, conf: SparkConf): String = {
+  def renameMainAppResource(
+      resource: String,
+      conf: Option[SparkConf] = None,
+      isJava: Boolean = true): String = {
     if (isLocalAndResolvable(resource)) {
-      SparkLauncher.NO_RESOURCE
+      if (isJava) {
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I think I'm missing some context, and others may be as well. Could we add a comment about why we need to upload the non-Java main resource?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: Maybe make that as a comment since most people won't go back to the PR to understand what is going.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 3 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I upload it because it resided at the client machine

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: So I guess the part I don't understand is why would it not need to be uploaded if it was Java? Presumably we upload the java somewhere else, why isn't the Python being uploaded in the same place?

**Review by holdenk — COMMENTED** [review]
Follow up question re: uploading.

**Comment by skonto** [comment]
@holdenk this is because spark-submit add the resource in sparks.jars:
```
19/10/11 13:01:30 WARN Utils: Your hostname, universe resolves to a loopback address: 127.0.1.1; using 192.168.2.4 instead (on interface wlp2s0)
19/10/11 13:01:30 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Parsed arguments:
  master                  k8s://https://10.96.0.1:443
  deployMode              cluster
  executorMemory          1G
  executorCores           null
  totalExecutorCores      null
  propertiesFile          null
  driverMemory            1G
  driverCores             null
  driverExtraClassPath    null
  driverExtraLibraryPath  null
  driverExtraJavaOptions  null
  supervise               false
  queue                   null
  numExecutors            2
  files                   null
  pyFiles                 null
  archives                null
  mainClass               org.apache.spark.examples.SparkPi
  primaryResource         local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar
  name                    spark-pi
  childArgs               [100]
  jars                    null
  packages                null
  packagesExclusions      null
  repositories            null
  verbose                 true

Spark properties used, including those specified through
 --conf and those from the properties file null:
  (spark.kubernetes.driver.pod.name,spark-pi-driver)
  (spark.executor.instances,2)
  (spark.driver.memory,1G)
  (spark.executor.memory,1G)
  (spark.kubernetes.authenticate.driver.serviceAccountName,spark-sa)
  (spark.kubernetes.namespace,spark)
  (spark.kubernetes.container.image,lightbend/spark:2.1.2-OpenShift-2.4.4-rh-2.12)
  (spark.kubernetes.container.image.pullPolicy,Always)

    
19/10/11 13:01:31 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Main class:
org.apache.spark.deploy.k8s.submit.KubernetesClientApplication
Arguments:
--primary-java-resource
local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar
--main-class
org.apache.spark.examples.SparkPi
--arg
100
Spark config:
(spark.kubernetes.namespace,spark)
(spark.jars,local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar)
(spark.app.name,spark-pi)
(spark.driver.memory,1G)
(spark.executor.instances,2)
(spark.submit.pyFiles,)
(spark.kubernetes.container.image.pullPolicy,Always)
(spark.kubernetes.container.image,lightbend/spark:2.1.2-OpenShift-2.4.4-rh-2.12)
(spark.submit.deployMode,cluster)
(spark.master,k8s://https://10.96.0.1:443)
(spark.kubernetes.authenticate.driver.serviceAccountName,spark-sa)
(spark.executor.memory,1G)
(spark.kubernetes.driver.pod.name,spark-pi-driver)
Classpath elements:
```
So since the PR [here](https://github.com/apache/spark/pull/23546) manages that property we dont need to do double work. For python its different.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 3 above)
Comment (erikerlandson) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: @holdenk does the java loading explaination work for you:
https://github.com/apache/spark/pull/25870#issuecomment-541001902

To confirm - the timing of the java loading is to ensure java code can execute when its needed?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 1 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: I added a comment above of each statement to clarify this.

**Comment by erikerlandson** [comment]
> @erikerlandson I added one test which uses a zip file as we discussed previously:

Thanks @skonto!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 3 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I think so, I was hoping we'd unify the code paths a bit more but that make sense.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 4]
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+      val cephUrlStr = getServiceUrl(svcName)
+      val cephUrl = new URL(cephUrlStr)
+      val cephHost = cephUrl.getHost
+      val cephPort = cephUrl.getPort
+      val examplesJar = Utils.getTestFileAbsolutePath(Utils.getExamplesJarName(), sparkHomeDir)
+
+      val (accessKey, secretKey) = getCephCredentials()
+      sparkAppConf
+        .set("spark.kubernetes.container.image", pyImage)
+        .set("spark.kubernetes.pyspark.pythonVersion", "2")
```
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: Lets not use Python 2 since this is targeted for Spark 3 I believe.

**Review by holdenk — COMMENTED** [review]
Minor comment about Python 2, but really excited to see this.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 4 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: I got used to 2 for years will miss it :) Sure will change.

**Comment by skonto** [comment]
@erikerlandson can I get a merge? Gentle ping :)

**Comment by skonto** [comment]
@holdenk gentl ping.

**Comment by skonto** [comment]
@dongjoon-hyun could you merge this one?

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @skonto . I'll take a look now.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 5]
```diff
@@ -259,12 +259,19 @@ private[spark] object KubernetesUtils extends Logging {
       isLocalDependency(Utils.resolveURI(resource))
   }
 
-  def renameMainAppResource(resource: String, conf: SparkConf): String = {
+  def renameMainAppResource(
+      resource: String,
+      conf: Option[SparkConf] = None,
+      isJava: Boolean = true): String = {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: Can we use `language-neural` parameter name, @skonto ? Then, this function can be reused for `SparkR` later.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala [hunk 6]
```diff
@@ -73,7 +76,9 @@ private[spark] class DriverCommandFeatureStep(conf: KubernetesDriverConf)
           .withValue(conf.get(PYSPARK_MAJOR_PYTHON_VERSION))
         .build())
 
-    val pythonContainer = baseDriverContainer(pod, res)
+    // re-write primary resource to be the remote one and upload the related file
+    val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), false)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala]: We recommend the following stye.
```scala
- val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), false)
+ val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), isJava = false)
```
The above is an example. When we rename `isJava`, we should change here together.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: Also, let's not have the default value. This function is used only twice. And you changed both places in this PR. There is no reason to introduce a new default value.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala [hunk 7]
```diff
@@ -170,4 +170,3 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
     additionalProps.toMap
   }
 }
-
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: I understand the intention, but let's not touch irrelevant part.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 8]
```diff
@@ -125,16 +125,18 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
       .endSpec()
       .build()
 
-    kubernetesTestComponents
+    // try until the service from a previous test is deleted
+    Eventually.eventually(TIMEOUT, INTERVAL) (kubernetesTestComponents
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: Is this relevant to this PR? If this aims to the flakiness of the existing test suite, we should make another PR before this one.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 9]
```diff
@@ -125,16 +125,18 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
       .kubernetesClient
       .services()
-      .create(cephService)
+      .create(cephService))
 
-    kubernetesTestComponents
+    // try until the stateful set of a previous test is deleted
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: ditto.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 10]
```diff
@@ -143,24 +145,26 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
       .statefulSets()
       .withName(cName)
+      .withGracePeriod(0)
       .delete()
 
     kubernetesTestComponents
       .kubernetesClient
       .services()
       .withName(svcName)
+      .withGracePeriod(0)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: ditto for the above two `.withGracePeriod(0)`s.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 11]
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
     }
   }
 
+  test("Launcher python client dependencies using py", k8sTestTag, MinikubeTag) {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: `using py`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 12]
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+        .set("spark.hadoop.fs.s3a.access.key", accessKey)
+        .set("spark.hadoop.fs.s3a.secret.key", secretKey)
+        .set("spark.hadoop.fs.s3a.connection.ssl.enabled", "false")
+        .set("spark.hadoop.fs.s3a.endpoint", s"$cephHost:$cephPort")
+        .set("spark.kubernetes.file.upload.path", s"s3a://$bucket")
+        .set("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+        .set("spark.jars.packages", "com.amazonaws:aws-java-sdk:" +
+          "1.7.4,org.apache.hadoop:hadoop-aws:2.7.6")
+        .set("spark.driver.extraJavaOptions", "-Divy.cache.dir=/tmp -Divy.home=/tmp")
+      createS3Bucket(accessKey, secretKey, cephUrlStr)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: +1 for refactor from line 204 ~ 224. We had better avoid this kind duplication.
As you suggested, please make this part as a utility function for the existing test and this new test.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala [hunk 13]
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
         s"under spark home test dir ${sparkHomeDir.toAbsolutePath}!")
     }
   }
+
+  def getTestFileAbsolutePath(fileName: String, sparkHomeDir: Path): String = {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: @skonto . In this PR, you copied `getExamplesJarAbsolutePath` to `getTestFileAbsolutePath` and removed all instances using `getExamplesJarAbsolutePath`. Do we think we need to maintain the unused function, `getExamplesJarAbsolutePath`?

In this case, I'd like to recommend you to generalize the existing function by renaming.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala [hunk 14]
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
...
+      zipOut.putNextEntry(zipEntry)
+      IOUtils.copy(fis, zipOut)
+      IOUtils.closeQuietly(fis)
+      IOUtils.closeQuietly(zipOut)
+    } catch {
+      case e: Exception => log.error(s"Failed to zip file: $inFile"); throw e
+    }
+  }
 }
+
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: In one file, you remove the empty line at the end.
- https://github.com/apache/spark/pull/25870/files#diff-7effa575514dd61752ece0b40fc2ad23L173
And here, you are adding a new empty line. 😄 .

Let's remove this redundant empty line.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala [hunk 15]
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
...
+      val fis = new FileInputStream(fileToZip)
+      val fos = new FileOutputStream(outFile)
+      val zipOut = new ZipOutputStream(fos)
+      val zipEntry = new ZipEntry(fileToZip.getName)
+      zipOut.putNextEntry(zipEntry)
+      IOUtils.copy(fis, zipOut)
+      IOUtils.closeQuietly(fis)
+      IOUtils.closeQuietly(zipOut)
+    } catch {
+      case e: Exception => log.error(s"Failed to zip file: $inFile"); throw e
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: Do we need to have `log.error` additionally when we re-throw the same `e`? Since the test case will fail eventually, I think we can skip the whole `try ... catch` effort.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @skonto . I finished my first round review. I'll review again after updating.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala (hunk 6 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala]: ok

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 7 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: ok

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 8 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: Nope this is to address waiting for the elimination of the running service from a previous test. It is an easy retry until the previous service is removed and the new can succeed.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala (hunk 13 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: Hmmm let me check.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 12 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: ok.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 11 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: yes correct.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 5 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: @dongjoon-hyun ok sure.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala (hunk 14 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: @dongjoon-hyun did I ok :) Usually I add a \n at the end as an old practice: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file
https://thoughtbot.com/blog/no-newline-at-end-of-file

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala (hunk 15 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: I added this so I can print the message for the failure. I can remove and see how informative the exception is.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala (hunk 10 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala]: We want to remove resources asap so grace period forces it, we dont want that?

**Comment by skonto** [comment]
Ok I will fix this over the weekend, sorry for the delay.

**Comment by skonto** [comment]
@dongjoon-hyun  could you re-open it pls? I will work on it.

**Comment by dongjoon-hyun** [comment]
Sure, no problem. Thanks, @skonto .

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala (hunk 13 above)
Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:137]: You are right we dont need it anymore.

Comment (skonto) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:137]: You are right we dont need it anymore, I will remove it.

**Comment by skonto** [comment]
@dongjoon-hyun ready for a review round. Refactored a few stuff. Thank you for your patience :)

**Comment by dongjoon-hyun** [comment]
@skonto . I visited this PR two days ago when you refreshed and visited now again.
But, my comments are still not addressed. That's the reason I didn't review this again.

**Comment by skonto** [comment]
@dongjoon-hyun sure will have a look.

**Comment by skonto** [comment]
@dongjoon-hyun I think addressed your comments, I missed a few earlier, had to scroll up a bit. There is a discussion about the use of `Eventually` but I have explained why I used this way to make sure the resources are cleaned up. Same for the use of the grace period there, it is no harm afaik.

**Comment by skonto** [comment]
@dongjoon-hyun gentle ping, pls let me know if I need to do anything more.

**Comment by dongjoon-hyun** [comment]
Very sorry for the delay. Since last Friday, I was very busy for the other stuffs. (The Jenkins outage and SparkSummit slide due). Since this is for 3.1.0, I'll try to re-visit here before this weekend. Hopefully, sooner than that.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by skonto** [comment]
@dongjoon-hyun thanks! I checked the failures but dont seem to be relevant. I will try re-test.

**Comment by skonto** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
The dependency issue is fixed.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
The job was running on `amp-jenkins-worker-05` which seems to have a corrupted Maven local info. This is another issue.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-cli) on project spark-parent_2.12: ArtifactInstallerException: Failed to install metadata org.apache.spark:spark-parent_2.12/maven-metadata.xml: Could not parse metadata /home/jenkins/.m2/repository/org/apache/spark/spark-parent_2.12/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got > (position: END_TAG seen ...</metadata>\n>... @13:2) -> [Help 1]
```

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
It's still assigned to `amp-jenkins-worker-05`.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by skonto** [comment]
Back to this.

**Comment by holdenk** [comment]
I'm also back after my motorcycle crash last year, let me know when it's up to date and I'd be happy to take another look. Thanks for sticking with this.

**Comment by erikerlandson** [comment]
I notice this is targeted for 3.1, is it worth trying to get on the 3.0.1 train?

**Comment by holdenk** [comment]
I think the plan is to cut 3.0.1 pretty soon since there are some almost show stopper bugs in 3.0 so I’d leave it targeted to 3.1 and once it’s in we can discuss if backporting makes sense.

**Comment by dongjoon-hyun** [comment]
+1 for @holdenk 's comment. Also, in general, we cannot backport the improvement JIRA.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @skonto . Apache Spark 3.1.0 Feature Freeze is scheduled on Early Nov 2020
- https://spark.apache.org/versioning-policy.html

**Comment by skonto** [comment]
@dongjoon-hyun I will try update, sorry it has been too long, my bad.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @skonto .

**Comment by skonto** [comment]
ack

**Comment by skonto** [comment]
@dongjoon-hyun I finally updated it pls review.

**Comment by dongjoon-hyun** [comment]
Thank you so much for updates, @skonto !

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala [hunk 16]
```diff
@@ -114,28 +116,39 @@ object Utils extends Logging {
...
         s"under spark home test dir ${sparkHomeDir.toAbsolutePath}!")
     }
   }
 
   def isHadoop3(): Boolean = {
     VersionInfo.getVersion.startsWith("3")
   }
+
+  def createZipFile(inFile: String, outFile: String): Unit = {
+      val fileToZip = new File(inFile)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:144]: indentation?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @skonto and all.
I'll fix the indentation and merge this.

Merged to master for Apache Spark 3.1.0.

**Comment by skonto** [comment]
Just saw your comment thanks!!!
